crates/amaru-ledger/src/rules.rs

```diff
@@ -0,0 +1,193 @@
+        .map_err(|err| BlockValidationError::RuleViolations(vec![err.into()]))?;
+    block_body_size_valid(&minted_block, &protocol_params)
+        .map_err(|err| BlockValidationError::RuleViolations(vec![err.into()]))?;
+
+    let ex_units = match block.transaction_witness_sets {
+        MaybeIndefArray::Def(vec) => vec,
+        MaybeIndefArray::Indef(vec) => vec,
+    }
+    .into_iter()
+    .flat_map(|witness_set| {
+        witness_set
+            .redeemer
+            .into_iter()
+            .map(|redeemers| match redeemers {
+                Redeemers::List(list) => list.iter().map(|r| r.ex_units).collect::<Vec<_>>(),
+                Redeemers::Map(map) => map.iter().map(|(_, r)| r.ex_units).collect::<Vec<_>>(),
+            })
+    })
+    .flatten()
+    .collect::<Vec<_>>();
+
+    block_ex_units_valid(ex_units, &protocol_params)
+        .map_err(|err| BlockValidationError::RuleViolations(vec![err.into()]))?;
+
+    // let transaction_bodies = block.transaction_bodies.to_vec();
+    // let transaction_witness_sets = block.transaction_witness_sets.to_vec();
+    // let invalid_transaction_indexes = block
+    //     .invalid_transactions
+    //     .map(|x| x.to_vec())
+    //     .unwrap_or(vec![]);
+    // for (i, body) in transaction_bodies.iter().enumerate() {
+    //     let is_valid = invalid_transaction_indexes.get(i).is_none();
+    //     let witness_set = transaction_witness_sets
+    //         .get(i)
+    //         .ok_or_else(|| BlockValidationError::SerializationError)?;
+
+    //     let state = validate_transaction(
+    //         body,
+    //         witness_set,
+    //         is_valid,
+    //         ledger_state,
+    //         protocol_params,
+    //         fast_fail,
+    //     )?;
+    // }
+
+    Ok((block_header_hash, minted_block))
+}
+
+// fn validate_transaction(
+//     body: &MintedTransactionBody,
+//     witness: &MintedWitnessSet,
+//     is_valid: bool,
+//     ledger_state: &Vec<usize>,
+//     protocol_params: &ProtocolParameters,
+//     fast_fail: bool,
+// ) -> Result<Vec<usize>, BlockValidationError> {
+//     // This was all scaffolding just as a PoC. Will implement real logic here
+//     // let utxos_exist = bad_inputs(10, &ledger_state);
+//     // let transaction_balanced = transaction_balanced(10, vec![10]);
+//     // let collateral_sufficient = collateral_sufficient(10, protocol_params, &ledger_state);
+
+//     // if let Err(error) = utxos_exist {
+//     //     if fast_fail {
+//     //         return Err(BlockValidationError::RuleViolations(vec![error.into()]));
+//     //     } else {
+//     //         return Err(BlockValidationError::Composite(
+//     //             error.into(),
+//     //             Box::new(vec![].into()),
+//     //         ));
+//     //     }
+//     // };
+
+//     todo!()
+// }
+
+fn parse_block(bytes: &[u8]) -> Result<(Hash<32>, MintedBlock<'_>), BlockValidationError> {
```

Comment:
```suggestion
#[instrument(level = Level::TRACE, skip(bytes), fields(block.size = bytes.len()))]
fn parse_block(bytes: &[u8]) -> Result<(Hash<32>, MintedBlock<'_>), BlockValidationError> {
```

I have all for moving code around but hey! Don't remove the traces!